### PR TITLE
Api updates

### PR DESCRIPTION
**Pull request by shermp**
Well, I've had a sudden spate of enthusiasm to work on this again.

There was a single breaking API change I wanted to make, so I figured if I'm breaking one thing, I may as well break 'em all. As far as I'm aware, Kobo-UNCaGED is currently the only consumer of this library, so I still have a chance.

Major changes are, in no particular order:
* Renaming some of the constants to Go naming conventions
* Error handling changes. More specifically, current idea is that a new error type has been added for calibre related errors. Client errors are returned without wrapping. All other errors (such as JSON, network) are wrapped. The client should use type assertion to check for UNCaGED specific errors
* Add the ability for UNCaGED to find multiple Calibre instances. The client interface has been extended to allow the client to choose from a list of found instances.
* Provide the SendBook() method with an io.Reader, instead of the client returning an io.WriteCloser. This means the client has to download the bytes from the TCP connection. On the other hand, the client has more control, especially if the client wishes to post-process the ebook.
* Some code refactoring

I'm also mulling the idea of creating a Calibre metadata parser, using the schema Calibre provides. That will be another PR though.

@geek1011 Would you mind having a look at this if you get a moment? I'm also open to any other suggestions for changes and improvements. May as well make them while I'm in the breaking mood.

### CODE REVIEW

**Comment by shermp** [comment]
Oops, meant to make this a draft PR... oh well.

**Comment by pgaskin** [comment]
From a quick glance, it looks good, but I'll do a full review and make some suggestions next week. Also, for error wrapping, remember that Go 1.13 has simple wrapping built in with fmt.Errorf and %w.

**Comment by shermp** [comment]
Thank you very much.

Error handling still isn't super great, but that seems to be the consequence of abstraction, at least in this case.

**Comment by shermp** [comment]
Right. Bit the bullet and implemented Go 1.13 style error wrapping. Hopefully I've made things a little more consistent and descriptive.

Only thing I'm a little unsure of is what to do with io.EOF. I've just left it as-is for now (not wrapping, and checking it using `==` method), The FAQ at https://github.com/golang/go/wiki/ErrorValueFAQ seems to indicate this is the correct approach.

**Comment by shermp** [comment]
Why did I not realise that `json.RawMessage` was a thing? Makes a better intermediary type compared to using `map[string]interface{}`. mapstructure is no longer required. It can all be done with `json.Unmarshal` now.

**Comment by shermp** [comment]
And one last major change. I...borrowed...the book metadata struct from Kobo UNCaGED, and modified it slightly. Should take the guesswork out of the top level metadata for clients. They no longer have to worry about what to do with a `map[string]interface{}`

I don't have any more changes planned. This PR has gotten big enough as it is (probably too big, but hey). It's now just a matter of implementing any fixes or cleanups before merging.

**Comment by shermp** [comment]
Sigh... somewhere in here, I've broken something with regards to `free_space_on_device`, and Calibre 4.x

Master still seems to work (at least on Kobo-UNCaGED it does), and it works with Calibre 3.48.

The problem as shown by wireshark is:
```[S]get_free_space [S]keep-alive noop [C]sends free space```
It SHOULD be:
```[S]get_free_space [C]sends free space [S]keep-alive noop```

*Sherman is confused*

EDIT: Yeah, definitely no issue with master. So I broke it somewhere in this PR :(

**Comment by shermp** [comment]
Umm... This is getting strange...

Started checking out different commits to try and narrow down where the problem might be. Somewhere Calibre stopped throwing an error, and I couldn't get it to show again. Revert back to my latest commit that was causing an error, and now no errors.

?!!!!?

*Sherman is feeling quite perplexed*

EDIT1: I can trigger the error again if the client has books to send metadata back to the server
EDIT2: Finally found the offending commit: 11d69499ef4b7e8693c940bbb2c1acdeb652b3ca

**Comment by shermp** [comment]
Arrgh... Really tearing my hair out over this one. I can't see anything I've done that would cause Calibre 4 issues. Especially when it's only Calibre 4 having issues :(

No doubt I've done something dumb somewhere that's in plain sight, and I just can't see it.

EDIT: I may or may not have redid the TCP handling to see if it was a timing issue. Needles to say, it didn't help. Not sure whether I'll keep the new version or not.

**Comment by shermp** [comment]
In Which Sherman Proves He Really Is A Dummy...

> No doubt I've done something dumb somewhere that's in plain sight, and I just can't see it.

Yes, yes that was indeed the case.
```
mdList := make([]uc.CalibreBookMeta, len(books))
	for _, bk := range books {
		for _, md := range cli.metadata {
			if bk.Lpath == md.Lpath {
				mdList = append(mdList, md)
			}
		}
	}
```
I was making `mdList` the correct size... then adding the actual metadata after it. Doh.

EDIT: It's not the first time this particular issue with slices has got me in trouble. When will I ever learn?

**Comment by shermp** [comment]
The PR that keeps on going. @geek1011 you need to stop me!

UNCaGED now requests an iterator from the client when getting full metadata from the client. I realised a simple slice might not be the best idea if the client needs to send hundreds or thousands of metadata items, especially if they include covers!

I think this is all the API changes... although that's what I said last time, so...

File: uc/uc.go [hunk 1]
```diff
@@ -61,34 +55,36 @@ func New(client Client, enableDebug bool) (*calConn, error) {
...
 	// five ports. We try all five ports concurrently
 	c.client.UpdateStatus(SearchingCalibre, -1)
 	bcastPorts := []int{54982, 48123, 39001, 44044, 59678}
+	wg := &sync.WaitGroup{}
+	mu := &sync.Mutex{}
 	for _, p := range bcastPorts {
-		go c.findCalibre(p, udpReply)
+		wg.Add(1)
+		go c.findCalibre(p, mu, wg)
 	}
```
Comment (pgaskin) [inline uc/uc.go]: This should be like this to follow the usual convention:

```go
	var wg sync.WaitGroup
	var mu sync.Mutex
	for _, p := range bcastPorts {
		wg.Add(1)
		go c.findCalibre(p, &mu, &wg)
	}
```

File: uncaged-cli/main.go [hunk 2]
```diff
@@ -42,29 +48,66 @@ type UncagedCLI struct {
 	bookDir      string
 	metadataFile string
 	drivinfoFile string
-	metadata     []map[string]interface{}
+	metadata     cliMeta
 	deviceInfo   uc.DeviceInfo
 }
 
+type cliMeta struct {
```
Comment (pgaskin) [inline uncaged-cli/main.go:55]: It might be a good idea to use a mutex here, just in case this is ever accessed concurrently.

**Review by pgaskin — COMMENTED** [review]
This is just a quick review from my phone. I'll take a more through look at the logic later (but note that I won't actually be testing it with a real Calibre instance, as I don't use Calibre). In general, it looks pretty good though.

File: uc/uc.go (hunk 1 above)
Comment (shermp) [reply uc/uc.go]: Took me a few moments to figure out what you had actually suggested, but yeah, have made that change.

File: uncaged-cli/main.go (hunk 2 above)
Comment (shermp) [reply uncaged-cli/main.go:55]: Hmm. Actually, looking at that code again, might be better to generate a one-time use iterator instead. That way multiple goroutines could iterate over the metadata independently without issues. Thanks for pointing that out.

Just as a side-note, the code in the cli path isn't really production quality. It's what I use to quickly test new features/changes in the library.

**Comment by shermp** [comment]
After all this time, I may as well merge this as-is. Improvements to the CLI code can be made at a later date, as the changes aren't related to the library API update.

**Comment by shermp** [comment]
Note, the changes made here should basically close #3 as well.

**Pull request merged**